src/celeritas/em/distribution/MuPPEnergyDistribution.hh

```diff
@@ -0,0 +1,249 @@
+    y_max_ = std::log(max_pair_energy_ / inc_energy_) / coeff_;
+
+    // Check that the bounds are within the grid bounds
+    CELER_ASSERT(y_min_ >= y_grid.front()
+                 || soft_equal(y_grid.front(), y_min_));
+    CELER_ASSERT(y_max_ <= y_grid.back() || soft_equal(y_grid.back(), y_max_));
+    y_min_ = max(y_min_, y_grid.front());
+    y_max_ = min(y_max_, y_grid.back());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the exiting pair energy.
+ */
+template<class Engine>
+CELER_FUNCTION auto MuPPEnergyDistribution::operator()(Engine& rng)
+    -> PairEnergy
+{
+    // Sample the energy transfer
+    real_type pair_energy
+        = inc_energy_ * std::exp(coeff_ * this->sample_scaled_energy(rng));
+    CELER_ASSERT(pair_energy >= min_energy_ && pair_energy <= max_pair_energy_);
+
+    // Sample the energy partition between the electron and positron
+    real_type r_max = (1
+                       - 6 * inc_mass_sq_
+                             / (total_energy_ * (total_energy_ - pair_energy)))
+                      * std::sqrt(1 - min_pair_energy_ / pair_energy);
+    real_type r = UniformRealDistribution(-r_max, r_max)(rng);
+
+    // Calculate the electron and positron energies
+    PairEnergy result;
+    real_type half_energy = pair_energy * real_type(0.5);
+    result.electron = Energy((1 - r) * half_energy - electron_mass_);
+    result.positron = Energy((1 + r) * half_energy - electron_mass_);
+
+    CELER_ENSURE(result.electron > zero_quantity());
+    CELER_ENSURE(result.positron > zero_quantity());
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the scaled energy and interpolate in log Z.
+ */
+template<class Engine>
+CELER_FUNCTION real_type
+MuPPEnergyDistribution::sample_scaled_energy(Engine& rng) const
+{
+    real_type u = generate_canonical(rng);
+    real_type y_lower = this->calc_scaled_energy(logz_interp_.index, u);
+    real_type y_upper = this->calc_scaled_energy(logz_interp_.index + 1, u);
+    return y_lower + (y_upper - y_lower) * logz_interp_.fraction;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the scaled energy for a given Z grid and sampled CDF value.
+ */
+CELER_FUNCTION real_type
+MuPPEnergyDistribution::calc_scaled_energy(size_type z_idx, real_type u) const
+{
+    CELER_EXPECT(z_idx < table_.grids.size());
+
+    TwodGridData const& cdf_grid = table_.grids[ItemId<TwodGridData>(z_idx)];
+    auto calc_cdf
+        = TwodGridCalculator(cdf_grid, table_.reals)(std::log(inc_energy_));
+
+    // Get the sampled CDF value between the y bounds
+    real_type cdf_min = calc_cdf(y_min_);
+    real_type cdf = std::fma(calc_cdf(y_max_) - cdf_min, u, cdf_min);
+
+    // Find the grid index of the sampled CDF value
+    NonuniformGrid y_grid(cdf_grid.y, table_.reals);
```

Comment:
I wonder if there's a nice way to abstract the rest of the function into a helper that takes `<class GridT, class CalcCdf>`... could make the code more testable and reusable?

Reply:
Done in #1525.